### PR TITLE
WIP: EBS Root Volume Termination

### PR DESCRIPTION
**Pull request by joshbranham**
I had a use case where setting the root EBS volume of nodes in an `InstanceGroup` to not delete on termination was useful. This setting was hardcoded, and thus root EBS volumes were always terminated. This is my first pass at enabling this as a configurable option in the `InstanceGroup` API. 

I implemented this by adding a new field to the `InstanceGroup` spec `retainVolumeOnTermination` per @justinsb recommendation. This field is then compared, and if set, the value for root volume termination is derived from it. Else, it's set to the default of always deleting the root volume when the instance is terminated.

Unsure if I need tests for this flow, but I have tested each possibility. I am also unsure if this needs to be implemented for the other block device mappings, or if we should leave it as root ebs only given my use case.

### CODE REVIEW

**Comment by joshbranham** [comment]
Functionally this is working, I am just having a bit of dificulty figuring out how to hide the diff output that is causing the tests to fail. Ideally, if you don't set the `RootVolumeTermination` field, the defaulting of `True` should be hidden from the end user. Therefore, if you set it to `False` it should show you that during `kops update cluster`. Any guidance would be helpful, thanks!

File: pkg/apis/kops/instancegroup.go [hunk 1]
```diff
@@ -103,6 +103,8 @@ type InstanceGroupSpec struct {
 	RootVolumeType *string `json:"rootVolumeType,omitempty"`
 	// If volume type is io1, then we need to specify the number of Iops.
 	RootVolumeIops *int32 `json:"rootVolumeIops,omitempty"`
+	// Keep the root volume upon instance termination
+	RootVolumeTermination *bool `json:"rootVolumeTermination,omitempty"`
```
Comment (justinsb) [inline pkg/apis/kops/instancegroup.go]: Naming suggestion: `retainVolumeOnTermination` or `deleteVolumeWhenInstanceTerminates` or something more self explanatory?

The `*bool` is good - it means that the flag will default to nil, and you don't have to make the existing behaviour  map to false - you can tell if someone explicitly set it or not

**Comment by justinsb** [comment]
Thanks for a great contribution @joshbranham 

The tests are failing because you need to "discover" the existing value for your new field from an existing ASG in AWS, in the method Find, I think similar to this:
https://github.com/kubernetes/kops/blob/16362e87a1ff8cbfec5221a3ac01716c5522d87c/upup/pkg/fi/cloudup/awstasks/launchconfiguration.go#L202

We discussed in [office-hours on 3/15](https://www.youtube.com/watch?v=SdW4Y7FgZfI&index=2&list=PL69nYSiGNLP2NcuGKoAzme6ZtZmrJlsaR&t=930s) the need to prioritize bug fixes until we can get 1.12 out (which will hopefully mean we get to everything faster overall), and I believe this is a feature, so I'm going to move this to the next release.
                                                                                                                                                                                                                           
/kind feature                                                                                                                                                                                                        
/milestone 1.13                                                                                                                                                                                                      
                   
I would love to see this in 1.13 - if the tests gets fixed up, and please squash your commits, and consider the naming suggestion - but otherwise looks great - nice work!

File: pkg/apis/kops/instancegroup.go (hunk 1 above)
Comment (joshbranham) [reply pkg/apis/kops/instancegroup.go]: @justinsb so with that in mind, the suggestion is to change the default volume termination to `false` if the field `retainVolumeOnTermination: true` exists?

**Comment by joshbranham** [comment]
Thank you for the info and pointers @justinsb, will check that out!

**Comment by kplimack** [comment]
@joshbranham I would love to use this feature, we have a similar use case.  Are you still planning on shepherding this pull-request through to master?

**Comment by joshbranham** [comment]
Our use case has fallen through and I was having some trouble getting existing functionality of the feature working with the field being exposed via my PR. It’s been not rebased for a while, but the basic idea is here, just needed more help as my golang is not great yet :)

**Comment by joshbranham** [comment]
Closing in favor of https://github.com/kubernetes/kops/pull/7865

**Pull request closed without merging**